druid/src/widget/immediate.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{widget::prelude::*, Data};
+
+/// A widget that allows writing parts of the ui in an immediate-mode style.
+///
+/// The primary usecase for this is displaying data that doesn't fit into druids model, such as
+/// enums you have no control over.
+///
+/// Whenever the state represented by the `Immediate` changes,
+/// it will reconstruct its content for the new data.
+///
+/// While this is not the most efficient thing to do, it is very simple and performs perfectly fine
+/// for small or rarely changed data.
+///
+/// Note that this is more like a last resort when the data to be represented is in a for druid particularly
+/// unfitting format.
+pub struct Immediate<D, W: Widget<()>> {
+    constructor: Box<dyn Fn(&D) -> W>,
```

Comment:
Perhaps it makes sense to have `Box<dyn Fn(&D) -> Option<W>>` for cases where we don't want to display any widget for events we don't care about?

Reply:
Hmm, I think I would prefer some kind of `Empty` widget that could be returned instead.
That could be useful in other places and allows keeping the api a bit simpler.

Reply:
Well as a developer using `Immediate` you could still return an `Empty` widget if you wish. I think the `Option` would enable a significant performance boost in certain cases. For example let's say you're only interested in displaying mouse click event info. Then every other event, especially events like size, will still cause an allocation of `Empty`. With `None` there would be no allocation.

Reply:
But if `Empty` is defined as
```rust
struct Empty;
```
It should not allocate anything, thus I kind of doubt that it would make a performance difference except having one more method call for each widget method :thinking:

Reply:
It's possible, but that way:
* Each developer needs to know about this even being a problem and possible to solve.
* Each developer needs their own bespoke implementation of `Empty`.

I think that common scenarios should be solved by the druid widget efficiently. Given that the whole idea here is to react to giant enums, not wanting to react to each and every variant seems rather common.

Reply:
My idea was that druid would provide such `Empty` widget, but I agree that it is a bit of a hidden solution and, on second thought, it would probably only be useful inside of `Immediate`.